tests/project/_internal/test_git_utils.py

```diff
@@ -24,17 +24,34 @@ def mock_progress():
 
 
 class TestClone:
-    def test_returns_repo(self, mock_progress, mock_repo):
+    def test_returns_repo(self, mock_progress, mock_repo, tmp_path):
         url = "https://blah"
-        path = Path()
+        path = Path(tmp_path, "dst")
         repo = git_utils.clone(url, path)
 
         assert repo is not None
         mock_repo.clone_from.assert_called_once_with(url, str(path), progress=mock_progress())
 
-    def test_raises_when_clone_fails(self):
-        with pytest.raises(VersionControlError):
-            git_utils.clone("", Path())
+    def test_raises_when_fails_due_to_bad_url(self, tmp_path):
+        with pytest.raises(VersionControlError, match="from url 'bad' failed"):
+            git_utils.clone("bad", Path(tmp_path, "dst"))
+
+    def test_raises_when_fails_due_to_existing_nonempty_dst_dir(self, mock_repo, tmp_path):
+        dst_dir = Path(tmp_path, "dst")
+        dst_dir.mkdir()
+        (dst_dir / "some_file.txt").touch()
+
+        with pytest.raises(VersionControlError, match="exists and is not an empty directory"):
+            git_utils.clone("https://blah", dst_dir)
+
+    def test_can_clone_to_empty_existing_dst_dir(self, mock_repo, tmp_path, mock_progress):
+        dst_dir = Path(tmp_path, "dst")
+        dst_dir.mkdir()
+        url = "https://repo"
+
+        git_utils.clone(url, dst_dir)
```

Comment:
Should we `assert repo is not None` or is this completely covered fine by `test_returns_repo()`? I see the test here is slightly different, in that we've created the dst_dir before calling clone. We might assume that the return from clone will not be None given no exceptions happened, but shouldn't we assert our assumptions? I understand this might be a bit of a stylistic choice. Let me know what you think best here.

Reply:
It is technically covered by `test_returns_repo`. However, we should really assert on the return value here, even if it's just for consistency with `test_returns_repo`, as these two test cases are verifying the "success case", so we should verify the postconditions in both cases.

Reply:
Asserted.